### PR TITLE
Perform add replica orchestrator calls outside coord thread

### LINKED ISSUE #15368: Change the replica creation API to return a `oneshot` which can be waited on in a separate thread
Currently the replica creation request blocks the main coord thread, meaning `SUBSCRIBE` and all other queries hang until the creation is complete (issue #14729). The solution is to move the create replica operation to its own thread.
A blocker for that is the ability to await a response from the replica creation API on a separate thread.

The proposed compute solution is to change the replica creation API to return a `oneshot`.

### PR DESCRIPTION
**Pull request by lluki**
This PR moves the orchestrator calls of `ActiveComputeController::add_replica_to_instance` and `ActiveComputeController::drop_replica` into the replica specific task. 

It should fix #15368, however not by returning a `oneshot` channel but rather by moving any network operation into the per replica tokio task.

### Tips for reviewers

 * First commit does add replica, second drop.
 * Note that `Instance::drop` was a no-op so far (see #14687). It now has the purpose of signaling the replica task to remove the service from the orchestrator and to terminate gracefully the background task.

### Motivation

  * This PR fixes a recognized bug. Fixes #15368

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note): None

### CODE REVIEW

**Comment by lluki** [comment]
Regarding the asynchronous replica removal: Assume we do the catalog transaction and asynchronously stop the replica in a background task. If computed crashes after the transaction, we get an orphaned computed. To fix this, `envd` could on startup determine orphans and remove them. 

```
used_computeds <- determine from catalog
all_services <- orchestrator.list_services()  
all_computeds <- filter all services by computed
for x in all_computeds - used_computeds:
   orchestrator.remove_service(x)
```

But I have to confirm this is correct orchestrator usage.

File: src/compute-client/src/controller/replicated.rs [hunk 1]
```diff
@@ -188,34 +189,45 @@ where
...
+                command = command_rx.recv() => match command {
+                    None => {
+                        // Controller is no longer interested in this replica. Shut down.
+                        return Ok(());
+                    }
+                    Some(mut command) => {
+                        if let ComputeCommand::DropInstance = command {
+                            // Controller is no longer interested in this replica. Shut down.
+                            return Ok(());
+                        };
```
Comment (teskje) [inline src/compute-client/src/controller/replicated.rs:205]: This code prevents the `DropInstance` command from being sent to the replica, which seems wrong.

File: src/compute-client/src/controller/replicated.rs [hunk 2]
```diff
@@ -77,27 +86,89 @@ where
...
+
+                        let addrs = service.addresses("controller");
+                        let comm_config = CommunicationConfig {
+                            workers: allocation.workers.get(),
+                            process: 0,
+                            addresses: service.addresses("compute"),
+                        };
+                        (addrs, comm_config)
+                    }
+                };
```
Comment (teskje) [inline src/compute-client/src/controller/replicated.rs:169]: Can we move the orchestrator call in a separate retry loop? If the connection to the replica fails and needs to be retried, there is no reason to also invoke the orchestrator again. I would also be fine with not retrying the orchestrator call at all and just panicking, but retrying is probably the safer default.

**Review by teskje — CHANGES_REQUESTED** [review]
Adding a replica looks good to me. I think we should do that in a retry loop that is separate from the one that tries to establish a connection to the replica though.

We should avoid dropping the replica in the orchestrator every time we need to rehydrate though. This seems very detrimental to performance and might introduce race conditions that could permanently wedge replicas.

**Comment by mjibson** [comment]
> To fix this, envd could on startup determine orphans and remove them.

I thought envd already did this?

File: src/adapter/src/coord/sequencer.rs [hunk 3]
```diff
@@ -1716,10 +1717,8 @@ impl<S: Append + 'static> Coordinator<S> {
         self.catalog_transact(Some(session), ops, |_| Ok(()))
             .await?;
         for (instance_id, replicas) in instance_replica_drop_sets {
-            for (replica_id, replica) in replicas {
-                self.drop_replica(instance_id, replica_id, replica.config)
-                    .await
-                    .unwrap();
+            for replica_id in replicas {
+                self.drop_replica(instance_id, replica_id).await?;
```
Comment (mjibson) [inline src/adapter/src/coord/sequencer.rs:1721]: Why does this get a `?` but we still use `.unwrap()` for `add_replica_to_instance`?

File: src/compute-client/src/controller.rs [hunk 4]
```diff
@@ -515,69 +524,21 @@ where
...
-            }
-        };
-
-        self.instance(instance_id)?.add_replica(
-            replica_id,
-            addrs,
-            config.logging,
-            communication_config,
-        )
+        self.instance(instance_id)?.add_replica(replica_id, config)
```
Comment (mjibson) [inline src/compute-client/src/controller.rs:533]: What's the state of the replica after this returns? Are any introspection queries that a users runs incorrect during the time when it's provisioning before it's ready? (Does ready even mean anything? I don't know.) Is "not yet ready" the same as "ready but not up to date yet" so it's the same to the user?

File: src/compute-client/src/controller/replicated.rs (hunk 1 above)
Comment (lluki) [reply src/compute-client/src/controller/replicated.rs]: I also fixed this, but right now we also swallow drop commands ;-)

File: src/compute-client/src/controller/replicated.rs (hunk 2 above)
Comment (lluki) [reply src/compute-client/src/controller/replicated.rs]: see above ^

File: src/compute-client/src/controller/replicated.rs [hunk 5]
```diff
@@ -188,34 +189,45 @@ where
...
+                    }
                 }
             }
         }
     }
+    .await;
+
+    if let ComputeReplicaLocation::Managed { .. } = location {
+        orchestrator.drop_replica(instance_id, replica_id).await?;
+    }
```
[the remark this answers is not in the document]
Comment (lluki) [reply src/compute-client/src/controller/replicated.rs]: Agree, I changed this now, the BG task will only remove from the orchestrator if it receives a DropInstance, otherwise it assumes that the reconciliation logic will restart the task.

**Comment by lluki** [comment]
> > To fix this, envd could on startup determine orphans and remove them.
> 
> I thought envd already did this?

Hm, I checked in catalog bootstrap/open/serve, and I can't find any orchestrator interaction there, but maybe i just missed that. I'll create a follow up issue to keep track of it (or just declare it resolved in case it's implemented already).

File: src/compute-client/src/controller.rs (hunk 4 above)
Comment (lluki) [reply src/compute-client/src/controller.rs:533]: It should be the same as before (apart from the orchestrator state). 

Before: after add_replica has been `.await`ed we were sure that the per replica catalog entries (introspection and system tables) are present in the catalog and the replica was registered with the orchestrator. But no guarantee about the replica being actually up or hydrated.

Now: After `add_replica` has finished (not async anymore) we can be sure that the per replica catalog entries are present. No guarantee anymore about the orchestrator state, but the background task will `ensure` this in the orchestrator until successful.

File: src/compute-client/src/controller/replicated.rs [hunk 6]
```diff
@@ -116,28 +171,58 @@ where
             // Command from controller to forward to replica.
             command = command_rx.recv() => match command {
                 None => {
-                    // Controller is no longer interested in this replica. Shut
-                    // down.
-                    return Ok(())
+                    // Controller is no longer interested in this replica. Shut down.
+                    bail!("command_rx received None")
```
Comment (lluki) [inline src/compute-client/src/controller/replicated.rs:175]: I made this case an error, now that we send correctly `DropInstance`, this should never happen.

**Comment by mjibson** [comment]
> Hm, I checked in catalog bootstrap/open/serve, and I can't find any orchestrator interaction there, but maybe i just missed that. I'll create a follow up issue to keep track of it (or just declare it resolved in case it's implemented already).

`ComputeCommand::InitializationComplete` is what I was talking about here (which you mentioned in slack).

**Comment by lluki** [comment]
> > Hm, I checked in catalog bootstrap/open/serve, and I can't find any orchestrator interaction there, but maybe i just missed that. I'll create a follow up issue to keep track of it (or just declare it resolved in case it's implemented already).
> 
> `ComputeCommand::InitializationComplete` is what I was talking about here (which you mentioned in slack).

This does the cleanup for all objects maintained in a running `computed` instance, so things like dataflows, indexes. But it does not interact with orchestrator or storage (and probably shouldnt, that's more an environmentd task).

File: src/compute-client/src/controller/replicated.rs [hunk 7]
```diff
@@ -642,17 +718,13 @@ where
         let replica_state = ReplicaState {
             command_tx,
             response_rx,
-            _task: task.abort_on_drop(),
```
Comment (teskje) [inline src/compute-client/src/controller/replicated.rs:645]: What is the reason to remove this? Seems like a good safe-guard to make sure we always clean up the task, even if we through some bug fail to do it otherwise.

**Review by teskje — COMMENTED** [review]
This looks good to me now, just a couple clarifying questions and nits. I would like to take another look after the rebase though, because the changes on main are substantial.

File: src/compute-client/src/controller/replicated.rs (hunk 7 above)
Comment (teskje) [reply src/compute-client/src/controller/replicated.rs]: Good point!

File: src/compute-client/src/controller/replicated.rs [hunk 8]
```diff
@@ -49,42 +48,97 @@ use crate::logging::LoggingConfig;
...
 
-async fn run_replica_core<T>(
-    ReplicaTaskConfig {
-        replica_id,
-        addrs,
-        build_info,
-        mut command_rx,
-        response_tx,
-    }: ReplicaTaskConfig<T>,
+/// Connects to replica (with a timeout) and runs message loop.
```
[the remark this answers is not in the document]
Comment (lluki) [reply src/compute-client/src/controller/replicated.rs]: Ah, we do retry forever, but with backoff I meant. Let me rephrase this

File: src/compute-client/src/controller.rs [hunk 9]
```diff
@@ -515,72 +524,22 @@ where
     ComputeGrpcClient: ComputeClient<T>,
 {
     /// Adds replicas of an instance.
-    pub async fn add_replica_to_instance(
+    pub fn add_replica_to_instance(
```
[the remark this answers is not in the document]
Comment (lluki) [reply src/compute-client/src/controller.rs:540]: I noticed after rebasing that it's not possible anymore to have `remove_replica` async, because of the same reason you just mentioned.
Unfortunately I think that `update_read_capabilities` can even cause a orchestrator interaction...
```
src/storage/src/controller.rs:1304:           hosts.deprovision(id).await?;
```

File: src/compute-client/src/controller/replicated.rs (hunk 6 above)
Comment (lluki) [reply src/compute-client/src/controller/replicated.rs]: :+1:

File: src/compute-client/src/controller.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (teskje) [reply src/compute-client/src/controller.rs:540]: Ah, I had hoped the await there would only do something short-blocking, but you are right, it talks to the orchestrator.

I think we should talk to the storage team to move the deprovisioning of hosts out of `update_read_capabilities`. That looks like it would make a lot of methods non-async in both controllers. It could happen similarly to how rehydration works now in the compute controller, i.e. `update_read_capabilities` doesn't deprovision directly but adds hosts-to-remove to some list, and then in the next call to `process` it drains that list and deprovisions all hosts inside.

File: src/compute-client/src/controller/instance.rs [hunk 10]
```diff
@@ -361,12 +371,23 @@ where
...
+            .add_replica(id, config.location, logging_config);
         Ok(())
     }
 
-    /// Remove an existing instance replica, by ID.
+    /// Remove an existing instance replica, by ID. Will remove the replica from the
+    /// orchestrator.
     pub async fn remove_replica(&mut self, id: ReplicaId) -> Result<(), ComputeError> {
+        if let Err(e) = self.compute.replicas[&id].send(ComputeCommand::DropInstance) {
+            tracing::warn!("Could not send DropInstance to replica {:?}: {}", &id, &e)
```
Comment (teskje) [inline src/compute-client/src/controller/instance.rs:382]: Since the orchestrator removal hinges on our ability to send `DropInstance` to the replica task, do you think it would make sense to elevate the logging level to error here? IMO, whenever we see this in the logs, we need to assume that we just leaked a computed pod, so we should go and clean it up manually. Maybe that would also be a good thing to add to the log message.

**Review by teskje — APPROVED** [review]
LGTM, if you can tweak that one log message a bit.

Have you tested this in a staging env a bit? Particularly it would be interesting to verify that we do the correct thing when a replica crashes at the same time (or close to the time) it is dropped.
